dts: bindings: Add `st,reinit-power-states` DT property

knthm opened the pull request:
### Summary

This PR adds the `st,reinit-power-states` DT binding property.
Said property is intended to describe exceptional System PM (sub)states, after which the defining peripheral instance requires reinitialization logic to regain normal operation.

Note: all states below and including S2RAM are expected to require driver action to restore operation, and these are not the target of this PR, but rather exceptions to the rule such as the S2idle state STOP2 on STM32WL.

A helper macro for building a `pm_state_info` struct array of these power-states, for various handling of critical driver sections is also included.

### Related

- #19755
- #59194 
- #37352
- #37414
- #59200

knthm committed:
dts: bindings: Add common st,power-management bindings

Add STM32-specific st,reinit-power-states property.

This property allows defining one or more special power-states phandles
in which a peripheral loses its configuration, and must be reinitialized.

States below and including S2RAM are inherently assumed to require
reinit, this property is intended to cover the exceptions outside
of this common use case.

It is intended to assist drivers in preventing the SoC from entering
these power states during critical code sections, as well as indicate
to power management services when a peripheral needs reinitialization.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented:
- Minor code compliance fixes

knthm commented:
- Minor code compliance and grammar fixes

knthm commented:
- Minor code compliance and grammar fixes
- Fix scenarios where `CONFIG_PM_DEVICE` might be defined without `CONFIG_PM`

knthm commented:
- Minor code compliance/typo/grammar fixes
- Fix scenarios where `CONFIG_PM_DEVICE` might be defined without `CONFIG_PM`

knthm commented:
- Minor code compliance/typo/grammar fixes
- Fix scenarios where `CONFIG_PM_DEVICE` might be defined without `CONFIG_PM`
- Replace `tests/benchmarks/footprints/benchmark.kernel.footprints.pm` test case for reinit with NULL - will fail due to missing `reinit-power-states` node property for select boards

gmarull commented on dts/bindings/base/pm.yaml:
```diff
@@ -30,3 +30,9 @@ properties:
       Automatically configure the device for runtime power management after the
       init function runs.
+
+  reinit-power-states:
+    type: phandles
+    description: |
+      System power states in which a peripheral's configuration is lost, and
+      must be reinitialized to ensure correct operation.
```
this has nothing to do with hw description, so it should use the `zephyr,` prefix.

knthm replied:
When it describes a (System PM) state in which the hardware changes, wouldn't that classify it as hardware description?

gmarull replied:
By looking at e.g. https://www.st.com/resource/en/product_training/STM32L4_System_Power.pdf, I see this property more of a software-driving property than something that precisely describes thee hardware.

erwango replied:
I agree in this case this is really a hw description, as it can be found in ref man
![wl_table](https://github.com/zephyrproject-rtos/zephyr/assets/18574308/3b7842c3-c89d-402a-9bfa-d9a7e493660d)

gmarull replied:
> I agree in this case this is really a hw description, as it can be found in ref man ![wl_table](https://user-images.githubusercontent.com/18574308/254269147-3b7842c3-c89d-402a-9bfa-d9a7e493660d.png)

I don't see LP states applying in the same way to all peripherals, so there seems to be a relationship between LP states and peripherals that this property doesn't capture.

knthm replied:
> I don't see LP states applying in the same way to all peripherals, so there seems to be a relationship between LP states and peripherals that this property doesn't capture.

That's why the property is intended to be used per-peripheral, e.g.

```c
		i2c2: i2c@40005800 {
                        ....
			reinit-power-states = <&stop2>;
                        ....
		};
```

Do you mean the wording or description of it doesn't capture the relationship? I could reword it if you think it's not clear enough.

knthm replied:
Note that the **Y**, **O**, and **R** classifications in the table above all mean the same thing in terms of peripheral configuration retention, if that is what you're touching on with your comment.
- **Y**: enabled, peripheral is operational during LP state
- **O**: optionally enabled (via config register bit) during LP state, but register/clock configuration is always retained
- **R**: register/clock configuration retained, though peripheral is unavailable during LP state

Only the lack of any classifier (denoted with a **-**) calls for peripheral reinit.
Adding a separate property to enable peripherals in LP states with an **O** would be useful, but is I believe out-of-scope for this PR's functionality.
Also note that the individual peripheral config bit to enable it applies to *all* LP states that have said **O** classifier, there's no option to enable for instance USART1 in STOP0, but not in STOP1, unless this is added to be done manually by System PM logic for every state exit/enter; I personally don't see a use-case for this.

ceolin replied:
This is a hw property for me as well.

gmarull commented on include/zephyr/pm/device.h:
```diff
@@ -120,6 +120,17 @@ typedef int (*pm_device_action_cb_t)(const struct device *dev,
 typedef bool (*pm_device_action_failed_cb_t)(const struct device *dev,
 					 int err);
 
+/**
+ * @brief PM device reinitialization callback.
+ *
+ * @param dev Device instance.
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP If the requested action is not supported.
+ * @retval Errno Other negative errno on failure.
+ */
+typedef int (*pm_device_reinit_cb_t)(const struct device *dev);
```
I think this can be solved in a simpler way:

Use the pm_notifiet_* API (now device agnostic, would require a device context in the callback), so that a driver gets called when the system enters/exits a power state. In these callbacks, driver can perform re-initialization if it knows that it has lost its state.

knthm replied:
That approach adds more overhead. For every `pm_state` exit, each driver will have to run comparison logic against the current power state and its list of `reinit-power-states`. I've had a similar discussion with @ceolin [here](https://github.com/zephyrproject-rtos/zephyr/pull/59200#discussion_r1263415999).

At least on the STM32 platforms, the SoC will regularly enter and exit low-power modes several dozen times a second, due to the short residency intervals. I think any additional overhead in this case should be avoided. (Adding calculation of exit-latency is another related case).

gmarull replied:
Can overhead be measured?

Assuming I have e.g. 2 devices subscribed, PM will just iterate over 2 list elements and call. Then, the callback will be something like

```
if (state == STOP2) {
   restore_context()
}

return;
```

Subscription could also be made per-state, maybe.

Before adding functionality such as "device reinit", I'd like to understand if this is something common, needed by other vendors to come up with something that is generic enough. For now this seems to be an STM32 requirement, so it can be solved inside drivers, not impacting other PM users.

knthm replied:
I suppose it adds a few cycles for each driver's notifier call, though it can quickly become a multiple of n, due to the nesting caused by substates (haven't tried to measure it directly). If you have 2 devices subscribed that require reinit in power-states `&stop0`, `&stop1`, and `&stop2`, you now have O(3n) complexity. This is the case on STM32WB for instance. I'm trying to think of and cover the worst case scenarios in my argumentation.

If you take quick look at [STM32WLE5x RM0461, chapter 5.4 - Low-Power Modes, Table 37](https://www.st.com/resource/en/reference_manual/rm0461-stm32wlex-advanced-armbased-32bit-mcus-with-subghz-radio-solution-stmicroelectronics.pdf), you'll see that there are **16(!)** peripherals that require reinit after STOP2 state. For other SoCs like STM32WB, this can start as early as STOP0, leading to a potential worst case of O(16*3) iterations to be run.

> Before adding functionality such as "device reinit", I'd like to understand if this is something common, needed by other vendors to come up with something that is generic enough. For now this seems to be an STM32 requirement, so it can be solved inside drivers, not impacting other PM users.

Very good point. It'd be good to know if this has a potential use case for other platforms. I'll do some cursory research to figure this out.

> Subscription could also be made per-state, maybe.

I also very much like this alternative if expanded to substates. It'd help reduce the O(m) complexity from per-driver-instance to O(m) in total. (where m = reinit substates)

knthm replied:
> I'm working on a proposal to simplify to vendor-specific states, ie, one single state holder. The current approach of aligning to ACPI lead to a quite cumbersome implementation

Very interesting! I agree, forcing MCUs into the ACPI sleep states pattern makes things unnecessarily difficult—especially when trying to fit the various flat MCU power state designs into substates under these.

knthm replied:
> Before adding functionality such as "device reinit", I'd like to understand if this is something common, needed by other vendors to come up with something that is generic enough. For now this seems to be an STM32 requirement, so it can be solved inside drivers, not impacting other PM users.

I've checked a few different vendor SoCs that are currently supported in Zephyr, and I've found at least one other that requires reinit functionality: the TI SimpleLink CC13x2/CC26x2 series.

According to the [reference manual](https://www.ti.com/lit/ug/swcu185f/swcu185f.pdf?ts=1689649773475), *chapter 7.6.4 - Standby Mode* (page 519):
*"All modules in MCU_VD with retention, as shown in Figure 7-3, are retained in standby mode. All other logic in MCU_VD must be reconfigured after wake up from standby mode."*
Modules in this case being TI-lingo for peripherals.

If you check *chapter 7.4 Digital Power Partitioning, Figure 7-3* (page 509), you'll see that this encompasses many generic peripherals such as I2C, SSI, etc.

Interestingly enough, the CC13xx/CC26xx I2C driver currently even does [reinit](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L294-L319) via [HAL](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L373-L375) in a manner very similar to the `pm_notify` callback implementation, but specific to the power state. *(It also does some [strange things](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L382-L384) related to power management I don't quite understand)*

Would this make you consider the implementation in this PR? I'd also create a separate PR for the appropriate changes to TI SimpleLink DTS and drivers affected. I still feel the DT property method is a bit more elegant than driver `pm_notify` callbacks.

ceolin replied:
> Regarding state/substate, I'm working on a proposal to simplify to vendor-specific states, ie, one single state holder. The current approach of aligning to ACPI lead to a quite cumbersome implementation, because MCUs have different requirements that in practice do not align or deviate too much.

note that MCUs are not the only one consuming it there are targets where ACPI is more aligned.

knthm replied:
It'd be nice to have coexistence between ACPI and custom MCU-specific power states. The substates solution is just clunky, IMHO.
After all, ACPI platforms are a minority on Zephyr (do correct me if I'm wrong).

ceolin replied:
> I suppose it adds a few cycles for each driver's notifier call, though it can quickly become a multiple of n, due to the nesting caused by substates (haven't tried to measure it directly). If you have 2 devices subscribed that require reinit in power-states `&stop0`, `&stop1`, and `&stop2`, you now have O(3n) complexity. This is the case on STM32WB for instance. I'm trying to think of and cover the worst case scenarios in my argumentation.
> 
> If you take quick look at [STM32WLE5x RM0461, chapter 5.4 - Low-Power Modes, Table 37](https://www.st.com/resource/en/reference_manual/rm0461-stm32wlex-advanced-armbased-32bit-mcus-with-subghz-radio-solution-stmicroelectronics.pdf), you'll see that there are **16(!)** peripherals that require reinit after STOP2 state. For other SoCs like STM32WB, this can start as early as STOP0, leading to a potential worst case of O(16*3) iterations to be run.
> 

That is not really the case right ? We don't transition from stop0 -> stop1 -> stop2, states change always from active to some of these power states. So it is not like you will be called 16  * 3 when going to sleep. The problem though is when you just need to reinit in stop2 but you will be called when going to to stop0 or stop1.

> > Before adding functionality such as "device reinit", I'd like to understand if this is something common, needed by other vendors to come up with something that is generic enough. For now this seems to be an STM32 requirement, so it can be solved inside drivers, not impacting other PM users.
> 
> Very good point. It'd be good to know if this has a potential use case for other platforms. I'll do some cursory research to figure this out.
> 
> > Subscription could also be made per-state, maybe.
> 
> I also very much like this alternative if expanded to substates. It'd help reduce the O(m) complexity from per-driver-instance to O(m) in total. (where m = reinit substates)

That is something that makes sense to me too.

ceolin replied:
The property makes sense to me, what I am not so sure is the additional callback, I would like to explore a little bit more the notifier path. It seems to me that we can make that work for this.

knthm replied:
> That is not really the case right ? So it is not like you will be called 16 * 3 when going to sleep

I meant in the case of a node having 3 substates in `reinit-power-states`, specifically with the pm_notifier callback only being aware of only the main `pm_state` (no substate-specific cb). In that case the driver callback comparison logic would have to check if the exit substate matches any of the three, thus in the worst case iterating three times before it hits a match on STOP2, for instance.

Also forgot to mentioned that of the 16 peripherals some have more than one instance, e.g. I2C-1 and -2, so worst-case would be more than 16 driver instance pm_notifier callbacks getting called (around 24ish).

If the two of you are still in favor of a pm_notifier approach over my approach in this PR:
Ideally we should modify the pm_notifier callback registration to behave more like the [TI SimpleLink HAL notifier registration](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_cc13xx_cc26xx.c#L373-L375), i.e. the callback gets called only when it matches the exact pm (sub)state entrance/exit.

Just give me the go-ahead that'd you'd both prefer this approach, and I'll look to implement it in place of the solution proposed in this PR.

knthm replied:
> Just give me the go-ahead that'd you'd both prefer this approach, and I'll look to implement it in place of the solution proposed in this PR.

> I would like to explore a little bit more the notifier path. It seems to me that we can make that work for this.

I'll take that as a green light from you :)

knthm commented:
- Add `pm_policy_state_lock_` helper macros—for preventing a driver from entering the SoC power-states forbidden by its instance's `reinit-power-states` during critical sections

<details>
<summary>Example usage</summary>

#### The added macros may be used in the following manner

```c
#include <zephyr/pm/policy.h>

static int i2c_stm32_transfer(const struct device *dev, ...)
{
        ...
	PM_POLICY_DT_DRV_INST_LOCK;
	ret = stm32_i2c_transaction(dev, *current, next_msg_flags, slave);
	PM_POLICY_DT_DRV_INST_UNLOCK;
        ...
}
```

Which, with the following node configuration
```c
i2c1: i2c@40005400 {
	reinit-power-states = <&stop1 &stop2>;
};
```

Will expand to
```c
static int i2c_stm32_transfer(const struct device *dev, ...)
{
        ...
        pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, 2);
        pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, 3);
	ret = stm32_i2c_transaction(dev, *current, next_msg_flags, slave);
        pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, 2);
        pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, 3);
        ...
}
```

knthm commented:
- Add `pm_policy_state_lock_` helper macros—for preventing a driver from entering the SoC power-states forbidden by its instance's `reinit-power-states` during critical sections

<details>
<summary>Example usage</summary>

#### The added macros may be used in the following manner

```c
#include <zephyr/pm/policy.h>

static int i2c_stm32_transfer(const struct device *dev, ...)
{
        ...
	PM_POLICY_DT_DRV_INST_LOCK;
	ret = stm32_i2c_transaction(dev, *current, next_msg_flags, slave);
	PM_POLICY_DT_DRV_INST_UNLOCK;
        ...
}
```

Which, with the following node configuration
```c
i2c1: i2c@40005400 {
	reinit-power-states = <&stop1 &stop2>;
};
```

Will expand to
```c
static int i2c_stm32_transfer(const struct device *dev, ...)
{
        ...
        pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, 2);
        pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, 3);
	ret = stm32_i2c_transaction(dev, *current, next_msg_flags, slave);
        pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, 2);
        pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, 3);
        ...
}
```

knthm commented:
- Add new reinit parameter to macro documentation in comments
- Retrigger now patched compliance checks (#60551)

ceolin commented:
> 

This is something that I liked, don't have concrete opinion about the implementation, but the functionality makes sense, it simplifies a lot the lock operation.

knthm commented:
@mbolivar-nordic @galak 

#### Regarding the supporting macros added in https://github.com/zephyrproject-rtos/zephyr/pull/60369/commits/98888b3518fe92a1110509a584d69b02622ac114:
Would it be better to place these in `include/devicetree.h` with reordered naming?
i.e. `DT_DRV_INST_PM_POLICY_REINIT_LOCK` instead of `PM_POLICY_INST_REINIT_LOCK`.

Since these are highly (PM Policy-) specific macros instead of generally applicable DT APIs, I've initially put them in `include/pm/policy.h`.

knthm commented on include/zephyr/pm/policy.h:
```diff
@@ -86,6 +86,51 @@ const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks);
+ */
+#define PM_POLICY_INST_REINIT_LOCK					\
+	DT_FOREACH_PROP_ELEM_SEP_VARGS(					\
+			DT_DRV_INST(0), reinit_power_states,		\
```
This is obviously problematic, we need a method to apply this to the correct instance.

knthm replied:
Solved via more central approach to the reinit state instance handling via driver config struct.

knthm committed:
devicetree: Add helper macros for STM32 pm states reinit

Add helper macros for building a pm_state_info array from a device's
st,reinit-power-states.

These macros are intended to assist drivers in and around device
reinitializiation.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented:
- Rebase on main
- Replace policy helper macros with generic helper for generating `pm_state_info` struct arrays from a device's `reinit-power-states`
  *Example usage to follow via separate PR*

knthm commented:
- Minor docs/comments updated to reflect the new state of this PR

erwango commented on dts/bindings/base/pm.yaml:
```diff
@@ -30,3 +30,9 @@ properties:
       Automatically configure the device for runtime power management after the
       init function runs.
+
+  reinit-power-states:
+    type: phandles
+    description: |
+      System power states in which a peripheral's configuration is lost, and
+      must be reinitialized to ensure correct operation.
```
I feel this should be limited to suspend-to-idle states.
For s2ram states, this should be considered as default.

knthm replied:
That'd require additional logic to handle and thus add complexity. Since the property describes hardware, I think it's a better idea to have it defined, even for states that may always require it, such as standby.

erwango replied:
Indeed, I understand the involved complexity, but I guess in most cases, there will be a state where all devices lose context by default, forcing to document it doesn't feel right.
One idea could be to make "standby" default, but this could be error prone for most users. Maybe we can add a cpu property for that if required if "standby" is not standard enough.
@ceolin, @gmarull thoughts?

knthm replied:
Adding on to this:
As per our discord discussion I agree it makes sense to have this property apply to non-s2ram (and deeper) states only.

This is following your suggestion of aligning with [the Linux definition](https://www.kernel.org/doc/html/next/admin-guide/pm/sleep-states.html#suspend-to-ram) of everything below (and including) the S2RAM inherently requiring reinitialization of peripherals:
> *In many cases, all peripheral buses lose power when entering S2RAM, so devices must be able to handle the transition back to the "on" state.*

I'll be updating the device tree definition description to align with this.

knthm commented:
- Rebase on main
- Update DT property description and docs:
  - Property exists as an corner-case exception for states above S2RAM

gmarull reviewed (commented):
I still fail to see why this property is needed. If any peripheral needs to be reinitialized when coming back from a particular power state, driver implementation will know this information and can subscribe to system power state changes without involving devicetree. Am I missing something?

knthm commented:
@gmarull As discussed [here](https://github.com/zephyrproject-rtos/zephyr/pull/60369#discussion_r1313815635), this property now applies to exceptions of the rule.
For example the STM32 standby low-power state: here all SoCs need to reinit, the drivers will register the applicable reinit logic without consideration of the `reinit-power-states` property.

Secondly, as discussed previously with you [here](https://github.com/zephyrproject-rtos/zephyr/pull/60369#discussion_r1267093807), the applicable states and affected peripheral instances vary.
How else are we going to specify in the driver that it should reinit for STOP2 on STM32WL for I2C-1,  but not 3, and not for most  other SoC series? This is what the devicetree is for, IMHO. Adding instance- and series-specific exceptions across all drivers is worse from a maintenance and readability perspective.

Thirdly, this property is responsible not just for reinit logic, but also preventing the driver from entering these states during critical sections via `pm_policy_state_lock_get/put`.

gmarull commented:
> @gmarull As discussed [here](https://github.com/zephyrproject-rtos/zephyr/pull/60369#discussion_r1313815635), this property now applies to exceptions of the rule. For example the STM32 standby low-power state: here all SoCs need to reinit, the drivers will register the applicable reinit logic without consideration of the `reinit-power-states` property.
> 
> Secondly, as discussed previously with you [here](https://github.com/zephyrproject-rtos/zephyr/pull/60369#discussion_r1267093807), the applicable states and affected peripheral instances vary. How else are we going to specify in the driver that it should reinit for STOP2 on STM32WL for I2C-1, but not 3, and not for most other SoC series? This is what the devicetree is for, IMHO. Adding instance- and series-specific exceptions across all drivers is worse from a maintenance and readability perspective.
> 
> Thirdly, this property is responsible not just for reinit logic, but also preventing the driver from entering these states during critical sections via `pm_policy_state_lock_get/put`.

Honestly, I'd make this a vendor-specific property for now, something like `st,reinit-power-states`, that is, do not involve the PM subsys for now because drivers can manage everything on their own. If we find other vendors end up following the same pattern, it can be generalized later (or not).

gmarull commented on include/zephyr/pm/state.h:
```diff
@@ -281,6 +295,60 @@ struct pm_state_info {
+ * Example devicetree fragment:
+ *
+ * @code{.dts}
+	soc {
+ *		...
+ *		cpu0: cpu@0 {
+ *			device_type = "cpu";
+ *			...
+ *			cpu-power-states = <&state0 &state1>;
+ *		};
+
+ *		i2c1: i2c@50000000 {
+ *			...
+ *			reinit-power-states = <&state1>;
+ *		};
+ *
+ *		power-states {
```
cpus/power-states do not belong to the soc (simple-bus) node, use `cpus`.

knthm replied:
Good catch, will update the docs.

knthm committed:
dts: stm32: Include st,power-management bindings where applicable

Add common STM32 power-management bindings to peripherals that require
the st,reinit-power-states property.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented:
@gmarull I've made this binding ST-specific. However the helper macros are still in `state.h`, I'm not sure it feels right to keep them there, even with a ST-specfic name. Do you have any suggestions for location where they can easily be used by the various different STM32 drivers?

gmarull commented on include/zephyr/pm/state.h:
```diff
@@ -182,6 +182,22 @@ struct pm_state_info {
 	COND_CODE_1(DT_NODE_HAS_STATUS(DT_PHANDLE_BY_IDX(node_id, cpu_power_states, i), okay),    \
 		    (PM_STATE_INFO_DT_INIT(DT_PHANDLE_BY_IDX(node_id, cpu_power_states, i)),), ())
 
+/**
+ * @brief Helper macro to initialize an entry of a struct pm_state_info array
+ * when using UTIL_LISTIFY in PM_STATE_INFO_LIST_FROM_DT_REINIT.
+ *
+ * @note Only enabled states are initialized.
+ *
+ * @param i UTIL_LISTIFY entry index.
+ * @param node_id A node identifier with compatible zephyr,power-state
+ */
+#define Z_PM_STATE_INFO_FROM_DT_REINIT(i, node_id)						  \
+	COND_CODE_1(DT_NODE_HAS_STATUS(DT_PHANDLE_BY_IDX(node_id, st_reinit_power_states, i),	  \
+				       okay),							  \
+		    (PM_STATE_INFO_DT_INIT(DT_PHANDLE_BY_IDX(node_id, st_reinit_power_states,	  \
+							     i)),),				  \
+		    ())
```
this should definitely be part of an internal ST header (something similar existed for pinmux before pinctrl existed, maybe you can check old releases like 2.6)

knthm replied:
Moved the macros to `include/zephyr/devicetree/pm_stm32.h`. I'd decided on `pm_stm32.h` as it may later be expanded with general STM32 pm exception cases, not just state-specific ones.

erwango reviewed (changes requested):
LGTM aside from the file used for headers.

erwango commented on include/zephyr/devicetree/pm_stm32.h:
```diff
@@ -0,0 +1,91 @@
+ * @param i UTIL_LISTIFY entry index.
+ * @param node_id A node identifier with compatible zephyr,power-state
+ */
+#define Z_PM_STATE_INFO_FROM_DT_REINIT(i, node_id)						   \
+	COND_CODE_1(										   \
+		DT_NODE_HAS_STATUS(DT_PHANDLE_BY_IDX(node_id, st_reinit_power_states, i), okay),   \
+		(PM_STATE_INFO_DT_INIT(DT_PHANDLE_BY_IDX(node_id, st_reinit_power_states, i)),),   \
+		())
```
I know this isn't exactly your current concern, but as we discussed, for 90-ish % of cases, reinit-power-state will be standby.
I wonder where we'll be able to get it taken into account:
I don't think this should be added in each peripheral driver since we'll just be duplicating information
Should/Could it be done directly in current macro ?

tmleman commented:
Why can't you use in this case PM_DEVICE_RUNTIME?

knthm commented:
@tmleman 
PM_DEVICE_RUNTIME actions aren't coupled to system PM events, except in very specific configurations. Also, on STM32 reinitialization needs to happen in a very specific point of SoC wakeup.
Please see the other linked PRs, they provide a clear case of why this is necessary and not covered by current PM functionality.
